src/app/doc_api.cpp

```diff
@@ -261,17 +261,46 @@ bool DocApi::cropCel(LayerImage* layer,
       newCelBounds.offset(-bounds.origin());
 
       gfx::Point paintPos(newCelBounds.x - newCelPos.x,
-                          newCelBounds.y - newCelPos.y);
+        newCelBounds.y - newCelPos.y);
 
+      const color_t bg = image->pixelFormat() == IMAGE_TILEMAP ?
+                           notile :
+                           m_document->bgColor(layer);
       newCelPos = newCelBounds.origin();
 
-      // Crop the image
+      doc::Grid grid;
+      if (layer->isTilemap()) {
+        const Tileset* tileset = static_cast<LayerTilemap*>(layer)->tileset();
+        grid = tileset->grid();
```

Comment:
This `grid` variable has `m_origin`=(0, 0). Then `grid.canvasToTile( some canvas bound )` will result in bounds (in tile units) referred to a grid  aligned with the canvas. So, `grid.canvasToTile( some canvas bound )` in the case which the tilemap has an offset of some pixels (>0 and less than one tilesize) will result in wrong tile bounds.

Reply:
Adjusting the `grid`'s origin here would affect the calculation of the `newPaintPos`. We only need the origin, one time, to get the right size of the new image. All the other calls to grid need the absolute positioning to work.

Reply:
I didn't read all these comments, but I wanted to say that it's important to set a correct grid origin (generally the tilemap cel position, as @Gasparoken said) so `grid.canvasToTile()` and `grid.tileToCanvas()` functions work as expected (in other case converting between canvas <-> grid/tiles coordinates will not work). All the code in Aseprite is programmed in such way, and that's the expected behavior.

For example, `grid.tileToCanvas(Point(0, 0))` should return the origin position of the grid (the tilemap cel position) instead of `Point(0, 0)`. And `grid.canvasToTiles(cel->position())` should return `Point(0, 0)`.

> Adjusting the grid's origin here would affect the calculation of the newPaintPos.

From the code `paintPos` is a delta in pixels (`newCelBounds.origin() - newCelPos`) and you are converting that delta to tiles (`newPaintPos = grid.canvasToTile(paintPos)`). I can see why in this case the grid should have origin=(0,0). My personal recommendation would be that you should try to avoid converting deltas in pixels directly with the grid (it will not work). Instead you should convert canvas positions (in pixels) to tiles, and then calculating the delta of tiles (maybe something like `grid.canvasToTile(newCelBounds.origin()) - grid.canvasToTile(newCelPos)`, but I'm not sure).